alpaca_trade_api/rest.py

```diff
@@ -269,10 +269,11 @@ def list_orders(self,
 
     def submit_order(self,
                      symbol: str,
-                     qty: int,
```

Comment:
when doing this (changing the order of arguments), it will break any previous usage of this package. many users and packages (e.g backtrader, pylivetrader) us it like this:
`api.submit_order("AAPL", 100, "buy", "market", "day")`
not saying we can't. just something to consider (and definitely document if we do decide to do that as a breaking change)

Reply:
@camelpac that's a good point. If we want to make `qty` optional while maintaining the current order of arguments, we could consider one of these approaches:

1. Make `qty` optional, leave `qty` in its current position, then add `notional` as the last kwarg. This maintains order at the expense of presenting the required `side`, `type`, and `time_in_force` arguments as optional.
2. Rename `qty` to something like `size`, then add `size_type` as the last kwarg, which accepts `"qty"` or `"notional"` and defaults to `"qty"`. This sidesteps the issue of making `side`, `type`, and `time_in_force` optional, at the expense of possibly breaking code that passes an explicitly named `qty` argument to `submit_order`.

What do you think?

Reply:
option 2 is not good if I understand you correctly.. you want to have both qty and size? that's confusing. 
option 1 is the right choice. BUT side, type and time_in_force are not "optional" they will just have a default value. optional implies they are not mandatory.. which is not correct. so it should look like this:

```py
def submit_order(self,
                 symbol: str,
                 qty: float = None,
                 side: str = "buy",
                 type: str = "market",
                 time_in_force: str = "day",
                 limit_price: str = None,
                 stop_price: str = None,
                 client_order_id: str = None,
                 extended_hours: bool = None,
                 order_class: str = None,
                 take_profit: dict = None,
                 stop_loss: dict = None,
                 trail_price: str = None,
                 trail_percent: str = None, 
                 notational: float = None)
```

Reply:
Thanks for your feedback @camelpac! To clarify, my concern about making `side`, `type`, and `time_in_force` _appear_  optional--even though they're not--was rooted in the mistaken assumption that we'd need to give them null default values to accommodate changes to `qty` without making `submit_order` too "opinionated" , so to speak. Since this seems to be a non-issue (i.e., there are reasonable default values for these arguments) I agree that Option 1 makes the most sense.

FWIW, this was what I had in mind for Option 2:

```py
def submit_order(
    self,
    symbol: str,
    size: float,  # abstracted from `qty`
    side: str,
    type: str,
    time_in_force: str,
    limit_price: str = None,
    stop_price: str = None,
    client_order_id: str = None,
    extended_hours: bool = None,
    order_class: str = None,
    take_profit: dict = None,
    stop_loss: dict = None,
    trail_price: str = None,
    trail_percent: str = None, 
    size_type: str = "qty",  # can be "qty" or "notional"
)
```
```

Reply:
Totally agree on the decision here for option 1. Another consideration would be to have a bail-out option to be exact the same behavior if one configures (e.g. environment variable) just in case. This is the first time breaking change on this most important method and it makes sense to pay extra coverage like that.

Reply:
> Another consideration would be to have a bail-out option to be exact the same behavior if one configures (e.g. environment variable) just in case.

@umitanuki did you have something like this in mind?

```py
try:
    resp = self.post('/orders', params)
    return self.response_wrapper(resp, Order)
except APIError as err:
    # ensure `qty` is non-null to avoid infinite recursion
    if (qty is not None) and os.environ.get("RESUBMIT_BAD_ORDERS"):
        # exclude notional parameter and resubmit order
        return self.submit_order(
            symbol,
            qty,
            side,
            type,
            time_in_force,
            limit_price,
            stop_price,
            client_order_id,
            extended_hours,
            order_class,
            take_profit,
            stop_loss,
            trail_price,
            trail_percent,
        )
    raise err
```

Reply:
option 1 looks good to me.. not sure what you mean by breaking change @umitanuki?

Reply:
I don't think introducing the env var makes sense because it still introduces a breaking change. better to just bump the minor version (instead of patch) since this doesn't break current usage but adds functionality